Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp

```diff
@@ -0,0 +1,217 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/Config/GeckoCodeWidget.h"
+
+#include <QFontDatabase>
+#include <QFormLayout>
+#include <QLabel>
+#include <QListWidget>
+#include <QMessageBox>
+#include <QPushButton>
+#include <QTextEdit>
+#include <QVBoxLayout>
+
+#include <QDebug>
+
+#include "Common/FileUtil.h"
+#include "Common/IniFile.h"
+#include "Core/ConfigManager.h"
+#include "Core/GeckoCodeConfig.h"
+#include "DolphinQt2/Config/CheatWarningWidget.h"
+#include "DolphinQt2/GameList/GameFile.h"
+
+GeckoCodeWidget::GeckoCodeWidget(const GameFile& game)
+    : m_game(game), m_game_id(game.GetGameID().toStdString()), m_game_revision(game.GetRevision())
+{
+  CreateWidgets();
+  ConnectWidgets();
+
+  auto game_ini_default = SConfig::GetInstance().LoadDefaultGameIni(m_game_id, m_game_revision);
+  IniFile game_ini_local;
+
+  // We don't use LoadLocalGameIni() here because user cheat codes that are installed via the UI
+  // will always be stored in GS/${GAMEID}.ini
+  game_ini_local.Load(File::GetUserPath(D_GAMESETTINGS_IDX) + m_game_id + ".ini");
+
+  m_gecko_codes = Gecko::LoadCodes(game_ini_default, game_ini_local);
+
+  UpdateList();
+}
+
+void GeckoCodeWidget::CreateWidgets()
+{
+  auto* layout = new QVBoxLayout;
+  m_warning = new CheatWarningWidget(m_game_id);
+  m_code_list = new QListWidget;
+  m_name_label = new QLabel;
+  m_creator_label = new QLabel;
+  m_code_description = new QTextEdit;
+  m_code_view = new QTextEdit;
+  m_download_codes = new QPushButton(tr("Download Codes (WiiRD Database)"));
+
+  QFont monospace(QFontDatabase::systemFont(QFontDatabase::FixedFont).family());
+
+  m_code_description->setFont(monospace);
+  m_code_view->setFont(monospace);
+
+  m_code_description->setReadOnly(true);
+  m_code_view->setReadOnly(true);
+
+  m_download_codes->setEnabled(!m_game_id.empty());
+
+  layout->addWidget(m_warning);
+  layout->addWidget(m_code_list);
+
+  auto* info_layout = new QFormLayout;
+
+  info_layout->addRow(tr("Name:"), m_name_label);
+  info_layout->addRow(tr("Creator:"), m_creator_label);
+  info_layout->addRow(tr("Description:"), static_cast<QWidget*>(nullptr));
+
+  info_layout->setFormAlignment(Qt::AlignLeft | Qt::AlignTop);
+
+  for (QLabel* label : {m_name_label, m_creator_label})
+  {
+    label->setTextInteractionFlags(Qt::TextSelectableByMouse);
+    label->setCursor(Qt::IBeamCursor);
+  }
+
+  const auto line_height = QFontMetrics(font()).lineSpacing();
+
+  m_code_description->setFixedHeight(line_height * 5);
+  m_code_view->setFixedHeight(line_height * 10);
+
+  m_download_codes->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
```

Comment:
These should be removed:
- We want `m_download_codes` to expand according to the height of the window, and this line gets ignored anyways (not sure why: maybe because it doesn't have a fixed size?).
- `m_code_description` and `m_code_view` already have their size policy effectively set to `(Expanding, Fixed)` by the calls to `setFixedHeight`.

Reply:
I agree about the second point (already removed it), but not on the first one: Why would we want the QPushButton to get more height or width?

Reply:
Oh, my mistake, since it was at the beginning of the group with `m_code_description` and `m_code_view` I thought it referred to the list of codes.

Actually, could you clean up this function a bit? Stuff that changes the same widget should be in the same place (e.g. `m_download_codes->setEnabled` and `m_download_codes->setSizePolicy`), and it'd also be nice if widgets were set up in the order that they're shown in the UI (since that's the natural order in which you think of them).